File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        HashMap <Character, Integer> count = new HashMap<>();
        int l=0;
        int maxFreq=0;
        int result=0;

        for(int r=0;r<s.length();r++){
            char rc=s.charAt(r);
            count.put(rc,count.getOrDefault(rc,0)+1);
            maxFreq=Math.max(maxFreq,count.get(s.charAt(r)));
            int replacements=(r-l+1)-maxFreq;
            if(replacements>k){
                char lc=s.charAt(l);
                count.put(lc,count.get(lc)-1);
                l++;
                replacements=(r-l+1)-maxFreq;
            }
            result=Math.max(result,r-l+1);
        }
        return result;
    }
}

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.py
class Solution(object):
    def characterReplacement(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        count={}
        left=0
        maxFreq=0
        result=0
        #add freqs to count
        for right in range(len(s)):
            count[s[right]]=count.get(s[right],0)+1

            #find most frequent character
            maxFreq=max(maxFreq,count[s[right]])

            #update replacements to windowSize-maxFreq
            replacements=(right-left+1)-maxFreq

            #if replacements exceed k, move window left
            while replacements>k:
                count[s[left]]-=1
                left+=1
                replacements=(right-left+1)-maxFreq

            #save largest window and return
            result=max(result,right-left+1)
        return result